File: programmers/level1/콜라츠 추측/solution.kt
fun solution(num: Int): Int {
    return calculateCollatzGuessCount(num.toLong(), 0)
}

fun task(num: Long) : Long = if (num % 2 == 0L) num / 2 else (num * 3) + 1

fun calculateCollatzGuessCount(num: Long, count: Int) : Int =
    when{
        count > 500 -> -1
        num == 1L -> count
        else -> calculateCollatzGuessCount(task(num), count+1)
    }

File: programmers/level1/콜라츠 추측/solution.js
/**
 * @param {number} num
 * @return {number}
 */
function solution(num) {
  return calculateCollatzGuessCount(num, 0);
}

/**
 * @param {number} num
 * @param {number} count
 * @returns {number} Collatz guess calculate count
 */
const calculateCollatzGuessCount = (num, count) =>
  count > 500
    ? -1
    : num === 1
    ? count
    : calculateCollatzGuessCount(task(num), count + 1);

/**
 * @param {number} num
 * @return {number} Collatz guess task
 */
const task = (num) => (num % 2 ? num * 3 + 1 : num / 2);
